x/wasm/internal/types/proposal.go

```diff
@@ -0,0 +1,263 @@
+		return sdkerrors.Wrap(govtypes.ErrInvalidProposalContent, "proposal title cannot be blank")
+	}
+	if len(p.Title) > govtypes.MaxTitleLength {
+		return sdkerrors.Wrapf(govtypes.ErrInvalidProposalContent, "proposal title is longer than max length of %d", govtypes.MaxTitleLength)
+	}
+
+	if len(p.Description) == 0 {
+		return sdkerrors.Wrap(govtypes.ErrInvalidProposalContent, "proposal description cannot be blank")
+	}
+	if len(p.Description) > govtypes.MaxDescriptionLength {
+		return sdkerrors.Wrapf(govtypes.ErrInvalidProposalContent, "proposal description is longer than max length of %d", govtypes.MaxDescriptionLength)
+	}
+	return nil
+}
+
+type StoreCodeProposal struct {
+	GovProposal
+	// Creator is the address that "owns" the code object
+	Creator sdk.AccAddress `json:"creator" yaml:"creator"`
+	// WASMByteCode can be raw or gzip compressed
+	WASMByteCode []byte `json:"wasm_byte_code" yaml:"wasm_byte_code"`
+	// Source is a valid absolute HTTPS URI to the contract's source code, optional
+	Source string `json:"source" yaml:"source"`
+	// Builder is a valid docker image name with tag, optional
+	Builder string `json:"builder" yaml:"builder"`
+}
+
+// ProposalType returns the type
+func (p StoreCodeProposal) ProposalType() string { return ProposalTypeStoreCode }
+
+// ValidateBasic validates the proposal
+func (p StoreCodeProposal) ValidateBasic() error {
+	if err := p.GovProposal.ValidateBasic(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// String implements the Stringer interface.
+func (p StoreCodeProposal) String() string {
+	var b strings.Builder
+
+	b.WriteString(fmt.Sprintf(`Store Code Proposal:
+  Title:       %s
+  Description: %s
+  Changes:
+`, p.Title, p.Description))
+	// todo: print all data
+	return b.String()
+}
+
+type InstantiateContractProposal struct {
+	GovProposal
+	// Creator is the address that pays the init funds
+	Creator sdk.AccAddress `json:"sender" yaml:"sender"`
+	// Admin is an optional address that can execute migrations
+	Admin     sdk.AccAddress  `json:"admin,omitempty" yaml:"admin"`
+	Code      uint64          `json:"code_id" yaml:"code_id"`
+	Label     string          `json:"label" yaml:"label"`
+	InitMsg   json.RawMessage `json:"init_msg" yaml:"init_msg"`
+	InitFunds sdk.Coins       `json:"init_funds" yaml:"init_funds"`
+}
+
+// ProposalType returns the type
+func (p InstantiateContractProposal) ProposalType() string {
+	return ProposalTypeStoreInstantiateContract
+}
+
+// ValidateBasic validates the proposal
+func (p InstantiateContractProposal) ValidateBasic() error {
+	if err := p.GovProposal.ValidateBasic(); err != nil {
+		return err
+	}
+	if err := sdk.VerifyAddressFormat(p.Creator); err != nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "creator is required")
+	}
+
+	if p.Code == 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "code_id is required")
+	}
+
+	if err := validateLabel(p.Label); err != nil {
+		return err
+	}
+
+	if p.InitFunds.IsAnyNegative() {
```

Comment:
I think there is a general "ValidateBasic" on sdk.Coins. Any reason not to use that? (It also checks denom and such)

Reply:
I did not know about this method. The code was copied from `MsgInstantiateContract`. I will change it there as well to `IsValid`
```go
	if !p.InitFunds.IsValid() {
		return sdkerrors.ErrInvalidCoins
	}
```